drv/stm32h7-eth/src/lib.rs

```diff
@@ -0,0 +1,485 @@
+                // Transmitted a packet.
+                packet_transmitted = true;
+                // Clear the interrupt. And the summary bit. Yes, we have to
+                // clear the summary bit, it is automatically set but not
+                // cleared.
+                self.dma.dmacsr.write(|w| w.nis().set_bit().ti().set_bit());
+            }
+        }
+        if dmaisr.macis().bit() {
+            // The MAC has an interrupt. We do not enable any MAC interrupts.
+        }
+        if dmaisr.mtlis().bit() {
+            // The MTL has an interrupt. We do not enable any MTL interrupts.
+        }
+
+        (packet_transmitted, packet_received)
+    }
+
+    /// Kicks off a SMI write of `value` to PHY address `phy`, register number
+    /// `register`. When this function returns, the write has started. It will
+    /// complete asynchronously, so feel free to do other things. Any attempt to
+    /// do other SMI operations will synchronize and block until the SMI unit is
+    /// free.
+    pub fn smi_write(&mut self, phy: u8, register: impl Into<u8>, value: u16) {
+        // Wait until peripheral is free.
+        crappy_spin_until(|| !self.is_smi_busy());
+
+        const WRITE: u8 = 0b01;
+
+        // Load data, then load address + start transaction. The address+start
+        // must be done using modify because the clock config is in the same
+        // register.
+        self.mac.macmdiodr.write(|w| unsafe { w.md().bits(value) });
+        self.mac.macmdioar.modify(|_, w| unsafe {
+            w.pa()
+                .bits(phy)
+                .rda()
+                .bits(register.into())
+                .goc()
+                .bits(WRITE)
+                .mb()
+                .set_bit()
+        });
+    }
+
+    /// Performs a SMI read from PHY address `phy`, register number `register`,
+    /// and returns the result.
+    ///
+    /// This function blocks until the read is complete, so that it can return
+    /// the result. It will also block until any previously issued write has
+    /// finished.
+    pub fn smi_read(&mut self, phy: u8, register: impl Into<u8>) -> u16 {
+        // Wait until peripheral is free.
+        crappy_spin_until(|| !self.is_smi_busy());
+
+        // Load address + start transaction
+        const READ: u8 = 0b11;
+        self.mac.macmdioar.modify(|_, w| unsafe {
+            w.pa()
+                .bits(phy)
+                .rda()
+                .bits(register.into())
+                .goc()
+                .bits(READ)
+                .mb()
+                .set_bit()
+        });
+
+        // Wait until it finishes.
+        crappy_spin_until(|| !self.is_smi_busy());
+
+        self.mac.macmdiodr.read().md().bits()
+    }
+
+    fn is_smi_busy(&self) -> bool {
+        self.mac.macmdioar.read().mb().bit()
+    }
+}
+
+/// Standard MDIO registers laid out in IEEE 802.3 standard clause 22. Vendors
+/// often add to this set in the 16+ range.
+pub enum SmiClause22Register {
```

Comment:
`vsc7448-pac` autogenerates a PHY register map based on Microchip's SDK, e.g. [`phy.rs` here](https://github.com/oxidecomputer/vsc7448/blob/master/vsc7448-pac/src/phy.rs) and the standard page [here](https://github.com/oxidecomputer/vsc7448/blob/master/vsc7448-pac/src/phy/standard.rs).

That's what I'm using for PHY bringup in `task-vsc7448`, since I need to use a bunch of registers from the extended page – does it make sense to also use it here?

Reply:
Maaaaaybe. That's a lot of code to bring in for boards that don't have a vsc7448 on them.

Reply:
After chatting at the Hubris sync, I'm going to make a separate `drv/vsc85xx` crate and use it in both the netstack and the `vsc7448` code.  That can wait until after this branch is merged, though.